pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/manager/SplitManager.java

```diff
@@ -0,0 +1,121 @@
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.pulsar.broker.loadbalance.extensions.channel.ServiceUnitState;
+import org.apache.pulsar.broker.loadbalance.extensions.channel.ServiceUnitStateData;
+import org.apache.pulsar.broker.loadbalance.extensions.models.SplitCounter;
+import org.apache.pulsar.broker.loadbalance.extensions.models.SplitDecision;
+
+/**
+ * Split manager.
+ */
+@Slf4j
+public class SplitManager implements StateChangeListener {
+
+    record InFlightSplitRequest(SplitDecision splitDecision, CompletableFuture<Void> future) {
+    }
+
+    private final Map<String, InFlightSplitRequest> inFlightSplitRequests;
+
+    private final SplitCounter counter;
+
+    public SplitManager(SplitCounter splitCounter) {
+        this.inFlightSplitRequests = new ConcurrentHashMap<>();
+        this.counter = splitCounter;
+    }
+
+    private void complete(String serviceUnit, Throwable ex) {
+        inFlightSplitRequests.computeIfPresent(serviceUnit, (__, inFlightSplitRequest) -> {
+            var future = inFlightSplitRequest.future;
+            if (!future.isDone()) {
+                if (ex != null) {
+                    counter.update(Failure, Unknown);
+                    future.completeExceptionally(ex);
+                    if (log.isDebugEnabled()) {
+                        log.debug("Complete exceptionally split bundle: {}", serviceUnit, ex);
+                    }
+                } else {
+                    counter.update(inFlightSplitRequest.splitDecision);
+                    future.complete(null);
+                    if (log.isDebugEnabled()) {
+                        log.debug("Complete split bundle: {}", serviceUnit);
+                    }
+                }
+            }
+            return null;
+        });
+    }
+
+    public CompletableFuture<Void> waitAsync(CompletableFuture<Void> eventPubFuture,
+                                             String bundle,
+                                             SplitDecision decision,
+                                             long timeout,
+                                             TimeUnit timeoutUnit) {
+
+        return eventPubFuture.thenCompose(__ -> inFlightSplitRequests.computeIfAbsent(bundle, ignore -> {
+            if (log.isDebugEnabled()) {
+                log.debug("Handle split bundle: {}, timeout: {} {}", bundle, timeout, timeoutUnit);
+            }
+            CompletableFuture<Void> future = new CompletableFuture<>();
+            future.orTimeout(timeout, timeoutUnit).whenComplete((v, ex) -> {
+                if (ex != null) {
+                    inFlightSplitRequests.remove(bundle);
+                    log.warn("Failed to wait for split for serviceUnit: {}", bundle, ex);
+                }
+            });
+            return new InFlightSplitRequest(decision, future);
+        }).future);
```

Comment:
We should update the counter here, because the `eventPubFuture` might be failed.